common/djangoapps/third_party_auth/api/views.py

```diff
@@ -89,3 +98,122 @@ def get(self, request, username):
+
+          GET /api/third_party_auth/v0/providers/{provider_id}/users
+
+          GET /api/third_party_auth/v0/providers/{provider_id}/users?usernames={username1},{username2}
+
+          GET /api/third_party_auth/v0/providers/{provider_id}/users?remote_ids={remote_id1},{remote_id2}
+
+          GET /api/third_party_auth/v0/providers/{provider_id}/users?usernames={username1}&remote_ids={remote_id1}
+
+    **URL Parameters**
+
+        * provider_id: The unique identifier of third_party_auth provider (e.g. "saml-ubc", "oa2-google", etc.
+          This is not the same thing as the backend_name.). (Optional/future: We may also want to allow
+          this to be an 'external domain' like 'ssl:MIT' so that this API can also search the legacy
+          ExternalAuthMap table used by Standford/MIT)
+
+    **Query Parameters**
+
+        * remote_ids: Optional. List of comma separated remote (third party) user IDs to filter the result set.
+          e.g. ?remote_ids=8721384623
+
+        * usernames: Optional. List of comma separated edX usernames to filter the result set.
+          e.g. ?usernames=bob123,jane456
+
+        * page, page_size: Optional. Used for paging the result set, especially when getting
+          an unfiltered list.
+
+    **Response Values**
+
+        If the request for information about the user is successful, an HTTP 200 "OK" response
+        is returned.
+
+        The HTTP 200 response has the following values:
+
+        * count: The number of mappings for the backend.
+
+        * next: The URI to the next page of the mappings.
+
+        * previous: The URI to the previous page of the mappings.
+
+        * num_pages: The number of pages listing the mappings.
+
+        * results:  A list of mappings returned. Each collection in the list
+          contains these fields.
+
+            * username: The edx username
+
+            * remote_id: The Id from third party auth provider
+    """
+    authentication_classes = (
+        OAuth2Authentication,
+    )
+
+    paginate_by = 50
+    paginate_by_param = 'page_size'
+    pagination_serializer_class = PaginationSerializer
+    serializer_class = serializers.UserMappingSerializer
+    provider = None
+
+    def get_queryset(self):
+        self.provider = Registry.get(self.kwargs.get('provider_id'))
+        if not self.provider:
+            raise Http404
+
+        # permission checking. We allow both API_KEY access and OAuth2 client credential access
+        if not (
+                ApiKeyHeaderPermission().has_permission(self.request, self) or
+                ThirdPartyAuthApiPermission().has_permission(self.request, self)
+        ):
+            raise exceptions.PermissionDenied()
+
+        usernames_str = self.request.QUERY_PARAMS.get('usernames', None)
+        remote_ids_str = self.request.QUERY_PARAMS.get('remote_ids', None)
+
+        query_set = UserSocialAuth.objects.select_related('user').filter(provider=self.provider.backend_name)
```

Comment:
@nasthagiri I see your point too, however I think you are missing a few pieces of the picture. The example you gave ("UniversityX" and "UniversityY") is currently impossible because I designed the OAuth provider configuration models to be uniquely keyed to the backend. So you can only ever have one "google-oauth2" provider configured on any given Open edX instance,. This is to avoid exactly the kind of problems you are worried about.

In addition:

* Due to the way the python-social-auth `UserSocialAuth` table works, the only way to implement multiple providers with the same backend is to use a `uid` prefix. (Or creating a whole new table to JOIN with but let's not go there.) So the current solution for querying the API for multiple providers is very generic. SAML uses `idp_slug` as the `uid prefix`, but other future implementations (such as a new sharable Google OAuth2 backend) could use some other key (such as a required domain name) as the `uid prefix`. Pan's code for this was perfectly generic and flexible. I believe that changing it to explicitly mention `idp_slug`s is making it less generic and thus less useful.

* I also am not in favour of making the admin SAML-only for now. As an example, right now, a school that uses Google Apps and has their own Open edX instance could configure the google provider (with the setting `WHITELISTED_DOMAINS = ["myschool.edu"]` which rejects other domains). With this API, they would then be able to use any student's Google Apps email address to lookup each student's edX account. But if you're making this SAML-only or SAML-specific for now, you're taking away that functionality.